### ISSUE TITLE
Add disconnection detection to pool

### ISSUE THREAD

**Issue by dvarrazzo** [psycopg/psycopg#247]
There is a [recipe to detect disconnections](https://www.psycopg.org/psycopg3/docs/advanced/async.html#detecting-disconnections) in the docs. It uses selectors and doesn't use polling, so it's pretty efficient.

There is some [advice about using it with the pool](https://www.psycopg.org/psycopg3/docs/advanced/pool.html#connection-quality) too. However it's up to the client to wrap up something to make use of it, and it wastes a connection entirely.

However we can make the extra step and add a disconnection detection feature? The pool is in a privileged position because it may sneakily use a connection internally, as long as it's well coordinated with external users. For instance, the recipe proposes:

```python
sel.register(conn, selectors.EVENT_READ)
while True:
    if not sel.select(timeout=60.0):
        continue  # No FD activity detected in one minute

    # Activity detected. Is the connection still ok?
    try:
        conn.execute("SELECT 1")
    except psycopg.OperationalError:
        await pool.check()
```

if we did this from within the pool, we could use a connection from within `pool._pool`, without taking it out of the available set. If activity is detected (so we are about to enter the `try` block), we may check `connection._pool`, which is populated only when the connection has been assigned to the client and taken out of the pool: in this case we would avoid to execute the `try` and start using a different connection instead.

Problems ahead:

- There are race conditions to take care of. We might want to check `conn._pool` holding the pool lock
- we should try to use the probing connection as last in the pool, to avoid to start a new selector every `min_size` requests (the pool is a queue: connection are taken with a popleft and returned with an append, because using it as a stack makes hard to implement `max_lifetime`). `pool._pool` is a deque, so inserting in penultimate position rather than append should be possible and efficient.
- if the probing connection is left out of the carousel then `max_lifetime` should be checked in a different way. So maybe it's worth to use a different probing connection every time, e.g. taking the most recent one and move it to the back to the deque, with a rotation...

**Comment by dvarrazzo** [comment]
`max_lifetime` is mostly used to avoid a connection to be opened _forever_ - you can see the reason for it from [this answer](https://www.postgresql.org/message-id/CAFj8pRCQN2B2vrVMH1-bd-8xtzjytWR%2BAjZ%2BMCj9J2wPxKPa9Q%40mail.gmail.com) from the PostgreSQL ML.

So, respecting `max_lifetime` to the second is not a hard requirement. Even if the pool has 1h max lifetime, and the pool stays unused for a month, there would be no activity on the connection to cause resource leakage (the connection rotting) and no client to be affected by it anyway.

`min_size=0` is quite an aggressive setting and can impact the performance of a pool much more than a connection remained open for longer than `max_lifetime`.

Is there a reason for which the pool should have a "contractual obligation" to respect the `max_lifetime` in a stricter way?

**Comment by aagmv** [comment]
>  Even if the pool has 1h max lifetime, and the pool stays unused for a month, there would be no activity on the connection to cause resource leakage (the connection rotting) and no client to be affected by it anyway.

The issue is what happens after the month. At that point there'll be a bunch of broken connections in the pool. With TCP keepalives it's quite likely that a connection is still usable after an hour. But after a month changing network conditions, server restarts and the like are nearly guaranteed to break it.

So it would be better if max_lifetime is also checked after popping from the pool and keep popping until a valid one is found or a new one is started.

**Comment by dvarrazzo** [comment]
Are we talking about a theoretical problem, or about something happening for real? In the latter case, on which parameters and timeframe does it happen?

And yes, if we get to implement a background check for connections, we can take in consideration that too.

**Comment by aagmv** [comment]
A real problem. It happened over the xmas holidays. No activity on the system and something in the network caused all TCP connections to the DB server to die but the pool didn't notice it. That in turn lead to the first users hitting the system getting errors because the number of dead connections exceeded the retry count.

It was only a few users and it recovered eventually, but it's still something I expected the `max_lifetime` setting to catch because those connections were way beyond their max.

**Comment by dvarrazzo** [comment]
The case of a prolonged pool inactivity is not really covered by `max_lifetime`.

A network disconnection can happen and can cause problems in a much shorter timeframe, well below max_lifetime: if the network goes down and you don't do anything about it, you will get minconn errors even one second later, not only after the holidays.

For this reason, an active disconnection detection, like explained in the first link in this ticket, or like what I'm thinking about building in this ticket, is the thing to do, rather than looping on the connections to enforce a strict respect for max_lifetime.

I'm also thinking that `check()` might be a good moment to check if `max_lifetime` is expired, because it's already in the business of taking out all the connections from the pool and exercise them. This is much simpler and can be done in a bugfix for psycopg_pool 3.1. It would be up to the user to call `pool.check()` periodically.

**Comment by dvarrazzo** [comment]
Added #482 to deal with it.

**Comment by aagmv** [comment]
It seems to be more natural to me to perform cheap checks while popping from the queue. It doesn't require the added complexity of having periodic tasks at all and it's intuitive because the max-lifetime is actually respected.

If you don't have a periodic check (which is the default) then max_lifetime... doesn't set a max_lifetime. That's just misleading.

> A network disconnection can happen and can cause problems in a much shorter timeframe

It can. But if you plot a cumulative probability distribution over the idle time then you'll see that it moves from 0 to 1, that's just the nature of cumulative probabilities. So having a short lifetime makes it _more likely_ that the connection is still alive without having to perform any active checks.

> rather than looping on the connections to enforce a strict respect for max_lifetime.

I didn't suggest any such thing. I suggested checking it one at a time while popping from the pool. If the pool is fresh enough you only check one connection. Amortized O(1). Only in the situation where everything has been idle for a long time there'll be some looping.

**Comment by dvarrazzo** [comment]
> It seems to be more natural to me to perform cheap checks while popping from the queue.

Define cheap check?

**Comment by aagmv** [comment]
The documentation says:

> The health of the connection is not checked when the pool gives it to a client.
> Why not? Because doing so would require an extra network roundtrip: we want to save you from its latency

Doing a numeric comparison (the connection age) after popping from a queue only takes a few CPU cycles and doesn't hit the network. In the unlikely event that it is too old there are still more connections in the pool to try. Only if all connections are beyond their max age it'll have to hit the network to create a new one.
But if one is at that point the whole system has been idle for so long that that would (at least in my case) have been the desired behavior.

**Comment by dvarrazzo** [comment]
During steady state, you don't need this check at all: the expired connection will be closed on putconn.

Being served a connection past max_lifetime is not a bad event per se. If there is a connection ready, and it's the last in the pool, discarding it instead of giving it to the client seems the wrong thing to do. My preference is to give the client the connection as it is, rather than make it pay the connection cost.

Your problem is that you are served a broken connection, not that you are served a connection past `max_lifetime`. Checking for a broken connection is not cheap: a network roundtrip, maybe even a rollback (unless we special-case `SELECT 1` to run in autocommit).

**Comment by dvarrazzo** [comment]
By the way, if you have:

- a long running program
- implementing no periodic health check
- which doesn't seem to need the low latency of a pool, if it can go totally unused for days

then maybe you would be better off using a `NullConnectionPool` instead.

**Comment by aagmv** [comment]
> During steady state, you don't need this check at all: the expired connection will be closed on putconn.

You asked me what's cheap. The fact that it's not needed most of the time doesn't change that it's cheap.
And steady-state, active operation may only occur during some times of the day. State transitions can happen often enough that they matter.

> Your problem is that you are served a broken connection, not that you are served a connection past max_lifetime.

As I explained above I understood the max-lifetime as a probabilistic approximation of a connection not being broken. If it worked 1 minute ago it is _more likely_ to work now rather than cases where it was last used 10 hours ago.

If that is not the purpose of `max_lifetime`, then what is its purpose?

> then maybe you would be better off using a NullConnectionPool instead.

My understanding is that that pool only caps the number of open connections, it doesn't reuse them at all.

I'd like to have a some pooling during activity spikes, but scaling down to zero is fine. Load is very uneven. You're right that I don't strictly need the low latency but reducing it when the system is in active use is a nice-to-have and may also increase throughput.

**Comment by dvarrazzo** [comment]
> If that is not the purpose of max_lifetime, then what is its purpose?

It was explained in a link above: to avoid bloat to accumulate in arbitrarily long connections. 

I am not interested in a probabilistic approach to connection quality check: that one can be easily achieved by a thread taking a connection every few seconds or an external agent asking the program to do the same. In my  long-lived programs I implement it typically using a k8s probe so that the system as a whole is informed about what is unhealthy. This goes beyond the psycopg scope of course.

> My understanding is that that pool only caps the number of open connections, it doesn't reuse them at all.

That's correct. It's a trade-off: it gives the same interface of a connection pool in case you want to use the same code with/without pooling behaviour. It's immune to problems caused by long-lived connections being interrupted, but doesn't save the client from the connection cost (unless there is a genuine spike, in which case the same connection is handed over from client to client). I use it too in some programs of mine, when there is a risk that connections get arbitrarily closed, and I'm ok with paying the connection price.

**Comment by dvarrazzo** [comment]
Thinking even more about it, if `max_lifetime` was used to check for broken connections, people would think it's a good idea for it to set it to one minute, whereas the idea about it is to have it in the order of the hours or days, for the reason explained above.

After `max_lifetime` has passed, the connection is closed unconditionally, whereas any "probabilistic" connection checker would need just to run a test query, not to close the connection. So it's really the wrong configuration knob to do that.

**Comment by retrojetpacks** [comment]
Maybe max lifetime needs a different name, or some additional description in the docs? I have also been making the assumption that max_lifetime was somehow checked while a connection was in the pool. The checking only on return to the pool does make sense and is consistent with my understanding of how psycopg works, but I only realised that after reading this thread!

**Comment by dvarrazzo** [comment]
#656 should be functionally similar to this idea, although more robust in the face of connections lost individually.

**Comment by cyounkins** [comment]
Isn't it possible to detect some connection closed conditions prior to using the connection? The server would send FIN or RST, and the client kernel would know the connection was closed. The driver could then check the socket for an error. If we can detect the closed socket and we're at the beginning of a transaction, we could switch to a new connection in the pool without risk.

If we can only detect disconnection after sending a query, would we be able to repeat queries on another connection if they fail due to disconnect?

I'm envisioning this would only occur at the start of transactions. It would be safe with read-only queries of course. If the query could modify state, I'm unsure if we could distinguish between retryable errors (such as `idle_session_timeout`) and non-retryable ones (server changed state, error was only in the acknowledgement). Thoughts?

**Comment by dvarrazzo** [comment]
The idea of detecting brokenness on a connection has some difficulties: we need to keep one connection aside for probing, and either not make it available to the client (essentially wasting it) or coordinating its usage with with clients usage. But the main problem I have seen is that it wouldn't guard from `idle_session_timeout` kicking in, which will happen at different times for different connections, or for some aggressive middleman such as the ones I'm seeing e.g. on Azure offerings of postgres-as-a-service.

These difficulties persist whether the probe is at tcp level or at application level.

**Comment by cyounkins** [comment]
Maybe I was not clear, or maybe there is a problem here that I'm not understanding. The solution I'm describing (checking if there is an error on the socket) does _not_ require a separate connection for probing, does not require a round trip delay, and _does_ guard against `idle_session_timeout`.

When a connection is closed by the server, a FIN or RST packet is sent from the server to the client. The client kernel is able to inform userland. `recv` would return -1, at which point `errno` could be checked.

So checking the socket status at the start of a transaction would catch many (but not all) connection closed conditions and allow for changing the connection to a valid one prior to sending the first statement.

**Comment by dvarrazzo** [comment]
@cyounkins I don't have a lot of experience with TCP/IP at that level. Are you able to put together a proof-of-concept?

**Comment by aagmv** [comment]
Note that you'll want to also use TCP keepalives to be able to detect a disconnect by just checking the local socket state. Otherwise some stateful middleboxes (nat, firewalls) may have forgotten about an idle connection and will blackhole all packets so the local machine doesn't learn about the severed connection until it tries to send packets over the network.

stateful gateways are also one of the reasons why time is a very good predictor whether a connection has or has not been closed. The probability transition from "works" to "doesn't" is quite sharp.

> When a connection is closed by the server, a FIN or RST packet is sent from the server to the client. The client kernel is able to inform userland. recv would return -1, at which point errno could be checked.

Combined with keep-alives a non-blocking `recv(..., MSG_PEEK)` is probably the most reliable way, but you can detect at least some disconnect states with `poll` and checking for pollerr/pollhup too.

> Thinking even more about it, if max_lifetime was used to check for broken connections, people would think it's a good idea for it to set it to one minute, whereas the idea about it is to have it in the order of the hours or days, for the reason explained above.

Well, one could distinguish between two different times. connection age and time-since-last-query.

**Comment by PaulVanSchayck** [comment]
We also came across a  similar situation. Maybe useful to know that it is possible to pass Postgress Connection Parameters Keywords via the `conninfo` string. 

This makes it possible to modify the TCP Keepalive settings for the connection, and also `tcp_user_timeout`. The latter governs  how long transmitted data may remain unacknowledged before a connection is forcibly closed. By default this is controlled (on Linux) by the sysctl `net.ipv4.tcp_retries2` setting, and is 15 minutes long. We've now set this to 15 seconds (15000 milliseconds).

Read more about these parameters here:

https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-PARAMKEYWORDS
